compiler/rustc_mir_transform/src/fold.rs

```diff
@@ -0,0 +1,190 @@
+use crate::MirPass;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+
+pub struct Fold;
+
+impl<'tcx> MirPass<'tcx> for Fold {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        while did_optimization(body, tcx) {}
+    }
+}
+
+fn did_optimization<'tcx>(body: &mut Body<'tcx>, tcx: TyCtxt<'tcx>) -> bool {
+    let mut visitor = AnalysisVisitor {
+        analysis: IndexVec::from_elem_n(Analysis::default(), body.local_decls.len()),
+    };
+    visitor.visit_body(body);
+    let analysis = visitor.analysis;
+
+    for result in analysis.iter() {
+        let write = result.write.as_assign(body);
+        let read = result.read.as_assign(body);
+        let Some(((temp_loc, (_temp_place, temp_rvalue)), (final_loc, (final_place, final_rvalue)))) = write.zip(read) else {
+                continue;
+            };
+        if temp_loc.block != final_loc.block {
+            continue;
+        }
+
+        /*
+        if let Rvalue::Cast(CastKind::IntToInt, _temp_op, _temp_ty) = temp_rvalue {
+            if let Rvalue::Cast(CastKind::IntToInt, _final_op, _final_ty) = final_rvalue {
+                //let initial_ty = write_op.ty(&body.local_decls, tcx);
+                //let intermediate_ty = write_ty;
+                //let final_ty = read_ty;
+
+                // Figure out how to prevent narrowing
+                //println!("{:?} {:?} {:?}", initial_ty, intermediate_ty, final_ty);
+                break;
+            }
+        }
+        */
+
+        if let Rvalue::Ref(_, _, place) = temp_rvalue {
+            if let Rvalue::AddressOf(mutability, final_operand) = final_rvalue {
+                let final_type =
+                    final_place.ty(&body.local_decls, tcx).ty.builtin_deref(true).unwrap().ty;
+
+                let mut new_projections = place.projection.to_vec();
+                for p in final_operand.projection.iter().skip(1) {
+                    new_projections.push(p);
+                }
+
+                let local = place.local;
+                let new_place =
+                    Place { local, projection: tcx.intern_place_elems(&new_projections[..]) };
+
+                assert_eq!(new_place.ty(&body.local_decls, tcx).ty, final_type);
+
+                let new = Rvalue::AddressOf(*mutability, new_place);
+                let new = StatementKind::Assign(Box::new((*final_place, new)));
+
+                let statements = &mut body.basic_blocks_mut()[temp_loc.block].statements;
+                statements[final_loc.statement_index].kind = new;
+                statements[temp_loc.statement_index].make_nop();
+
+                expand_live_range(body, final_loc.block, local);
+
+                return true;
+            }
+        }
```

Comment:
This optimisation is wrong for mutable addresses.
You may have something like:
```
_1 = &mut x
_2 = &raw mut *_1
_3 = *_2
_4 = *_1
```

You optimization turns this into:
```
_1 = &mut x
_2 = &raw mut x // uses `x`, so `_1` is cannot be used any more
_3 = *_2
_4 = *_1 // UB
```

Reply:
This case should be avoided because `_1` is read from twice. `Analysis` for the local `_1` should have `read: Status::Multiple`.

Reply:
Another issue that surfaced in https://github.com/rust-lang/rust/pull/106285: `&raw (*_1)` should never be UB, but `&raw _2` may be UB if `_2` is storage-dead at the point where the `AddressOf` statement appears.

More generally, this block appears like a restricted version of https://github.com/rust-lang/rust/pull/106285. Is there a case where it's broader in scope?

Reply:
Correct, I believe this does a weaker form of your SSA analysis. The upside is that this analysis is enormously faster to run.